### PR TITLE
Add option for background thread reporting

### LINKED ISSUE #124: Background Coverage Data Reporting
We're trying out coverband on production, and finding that the redis calls are taking a majority of the time of some of our web requests when looking at new relic. 

Digging into the redis_store, it seems that, for every file, we first get all of the existing values using `existing = redis.hgetall(key)`, then after incrementing the line counts, we use `redis.mapped_hmset(key, values)` to update redis.

In our application this results in hundreds of gets, and hundreds of sets to redis in the middleware. I have a couple ideas and I'm curious what the devs here think:

1. Fork this call, https://github.com/danmayer/coverband/blob/master/lib/coverband/middleware.rb#L14 . The web request doesn't need to wait for anything from this report generation, so why not let it run in another process?

2. Pipeline the calls instead of doing things one file at a time. In our case, that would mean 2 redis commands instead of ~300 on some requests. 

ie, here: https://github.com/danmayer/coverband/blob/master/lib/coverband/adapters/redis_store.rb#L26 ,  we iterate through every file in the report, and for each we get the existing values then increment and update here: https://github.com/danmayer/coverband/blob/master/lib/coverband/adapters/redis_store.rb#L53 . It looks like redis supports pipelining, https://redis.io/topics/pipelining, which would allow all these requests to be grouped together.  I've tried this and it seems to work:

```ruby
def save_report(report)
  store_array(base_key, report.keys)

  keys = report.keys.map {|file| "#{base_key}.#{file}"}

  existing_records = redis.pipelined do
    keys.each do |key|
      redis.hgetall(key)
    end
  end

  combined_report = {}
  keys.each_with_index do | key, i|
    combined_report[key] = {
      new: report.values[i],
      existing: existing_records[i]
    }
  end

  redis.pipelined do
    combined_report.each do |file, values|
      existing = values[:existing]
      new = values[:new]
      unless values.empty?
        # in redis all keys are strings
        new_string_values = Hash[new.map { |k, val| [k.to_s, val] }]
        new_string_values.merge!(existing) { |_k, old_v, new_v| old_v.to_i + new_v.to_i }
        redis.mapped_hmset(file, new_string_values)
        redis.expire(file, @ttl) if @ttl
      end
    end
  end

  # report.each do |file, lines|
  #   store_map("#{base_key}.#{file}", lines)
  # end
end
```

(extra credit); not sure if redis support this, but perhaps the update could be done with one command? Could you tell redis 'here are all the keys (files) and their values (arrays of line# counts), and if you find existing values for those keys, increment by the count given.

### PR DESCRIPTION
**Pull request by kbaum**
resolves #124 

Kicks off thread from the middleware on first request `if Configuration#backround_reporting_enabled`.

Currently only handles background reporting from web application servers through middleware.

TODO: Kick off background thread from `Coverband.start` for other ruby processes like sidekiq.

### CODE REVIEW

File: coverband.gemspec [hunk 1]
```diff
@@ -28,6 +28,8 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'test-unit'
   spec.add_development_dependency 'redis'
   spec.add_development_dependency 'benchmark-ips'
+  spec.add_development_dependency 'm'
```
[the remark this answers is not in the document]
Comment (danmayer) [reply coverband.gemspec]: cool thanks, I have wanted that feature but been to lazy to set it up ;)

**Comment by danmayer** [comment]
Nice looks good @kbaum considering background jobs sidekiq and cron already require some custom code to support, do we think this would be good enough for the first usable version for 3.1... I am also going to try to think of a way to benchmark this, I am pretty sure enabling this on coverband_demo and running the apache benchmark tests against it would show an improvement.

**Comment by kbaum** [comment]
I agree that this is probably good enough for 3.1 considering sidekiq already requires custom code. That being said, I think with this addition, sidekiq and other non-rack ruby processes would work well:
https://github.com/danmayer/coverband/commit/1067fab3b342e1604dd55409868ad3b0d1b0bae0
The idea is that if it's a rack server, the thread gets kicked off on the first web request. All other processes like sidekiq, delayed_job, plain scripts, rake tasks, etc kick off the background reporter on Coverband.start. 

On another note, I noticed this morning that what i had pushed last night was not working since the code was actually referencing Coverband::BackgroundReporter which I had renamed to Coverband::Background. First obviously I didn't do enough manual testing but this is also telling me that we are lacking some test coverage. I looked at [ collectors_coverage_test](https://github.com/danmayer/coverband/blob/master/test/unit/collectors_coverage_test.rb) and it seems to be missing coverage on report_coverage. I added a basic test that tests the changes in this PR, but I'm wondering if we should add some coverage in this area before cutting a new release.

Let me know what you think.
thx.

**Comment by danmayer** [comment]
yeah agreed we are missing some coverage, I think tests for that got lost and removed in the refactoring pushing everything back into a single collector... I created this issue to try to make it a bit easier to see when test coverage accidentally gets reduced https://github.com/danmayer/coverband/issues/136

We do probably need to get some coverage back on that method, although, I do think a coverage report would be helpful as it likely does get tested in the full stack test: https://github.com/danmayer/coverband/blob/master/test/unit/full_stack_test.rb

probably need to do a pass on the test folder making integration, unit, and other tests a bit more clear.

**Comment by danmayer** [comment]
makes sense @kbaum we could reraise run in test mode, I guess just need to figure out what to check for being in that mode. So used to `Rails.env`, but guessing minitest has some environment var we can check.

**Comment by kbaum** [comment]
Think we could just set a configuration in teat_helper.rb.

**Pull request merged**